Notify user on search error

DadImScared opened the pull request:
https://trello.com/c/rPWmb1mh/130-notify-user-on-search-error

DadImScared committed:
Show error on search error

anantakrishna commented on frontend/src/search.js:
```diff
@@ -83,6 +83,34 @@ search.addWidgets([
+        document.getElementById('error').classList.add('hidden');
+        this.message = '';
+        setTimeout(() => {
+          search.helper.search();
```
Here and below we are referring the global variable `search`, which makes the widget not isolated. Can't we get the `helper` reference from the parameters of the `init` method or any other way natural for a widget?

DadImScared replied:
[Documentation](https://www.algolia.com/doc/guides/building-search-ui/going-further/conditional-display/js/#handling-errors) shows we can access the helper in the `init` method but hooking into the error event through the helper does not work as expected.

anantakrishna replied:
Very relevant link! I regret I had not found it in the first place. It's exactly about what we want to achieve — error-handling widget.

Can you explain in details what you mean by “does not work as expected”? How does it work?

anantakrishna replied:
Thanks, I confirm that the error handler does not get triggered. I've posted a question on the official forum: https://discourse.algolia.com/t/error-handling-widget/13384. Let's wait for a day or two.

anantakrishna commented on frontend/src/search.js:
```diff
@@ -83,6 +83,34 @@ search.addWidgets([
+      });
+      search.on('error', ({ error }) => {
+        this.message = error.message;
+        document.getElementById('error-message').innerHTML = error.message;
```
Why wouldn't we do this in the `render` method?

anantakrishna commented on frontend/src/search.js:
```diff
@@ -83,6 +83,34 @@ search.addWidgets([
+          search.helper.search();
+        }, 500);
+      });
+      search.on('error', ({ error }) => {
```
I doubt we need to hook this event within a widget. Please see if we can get all the necessary information from the `searchMetadata` in the `render` method.

DadImScared replied:
`searchMetadata` does not include any information about an error.

anantakrishna replied:
This discussion is related to the [other thread](https://github.com/PureBhaktiArchive/archive-web-app/pull/120/files/b69b15307729d874105b650bc48fff86bdb0361b#r683480695). Let's proceed there.

anantakrishna commented on frontend/src/search.js:
```diff
@@ -83,6 +83,34 @@ search.addWidgets([
+    },
+    render({ searchMetadata = {} }) {
+      const { isSearchStalled } = searchMetadata;
+      document
+        .getElementById('stats')
+        .classList.toggle('hidden', this.message || isSearchStalled);
```
Let's try top avoid manipulating `stats` widget from here and from the `loading` widget as well. We can do this by customising the `stats` widget itself: https://www.algolia.com/doc/api-reference/widgets/stats/js/#connector. It can react on the error/stalled states and hide itself.

anantakrishna commented on frontend/src/search.js:
```diff
@@ -83,6 +83,34 @@ search.addWidgets([
     },
   },
+  {
+    message: '',
```
I'd prefer `null` to indicate the absence of error instead of an empty string.

DadImScared committed:
Remove algolia error

DadImScared committed:
Change retry button to link

DadImScared committed:
Use helper for error event

northern-mindset committed:
Merge branch 'main'

Commit:
removed old code for error notification.

Commit:
updated element for error message

Commit:
updated error widget code

Commit:
status widget created with different search status notification

northern-mindset committed:
Don't manipulate irrelevant elements from widget

northern-mindset committed:
Implement the retry logic using `UIState`

northern-mindset committed:
Refactor HTML and CSS of the indicators

Commit:
search-bar.js file renamed to search-box.js

Commit:
renamed searchBar component to searchBox

Commit:
updated search loading spinner

Joydev commented on frontend/src/audios.html:
```diff
@@ -81,12 +93,15 @@
-      <span id="loading" class="text-sm font-bold italic text-neutral-600">Loading…</span>
+      <div class="text-sm text-neutral-600">
+        <span id="stats" class="whitespace-nowrap"></span>
+        <span id="loading-indicator" class="">Loading…</span>
```
previous loading indicator removed.

Commit:
removed old loading indicator element.

Joydev commented on frontend/src/audios.html:
```diff
@@ -61,12 +61,24 @@
         class="block w-full appearance-none rounded-none rounded-l-md border border-r-0 border-gray-300 px-3 py-2 leading-6 placeholder-gray-300 focus:border-gray-300 focus:outline-none focus:ring-0"
       >
-      <button type="submit" class="sm:px4 rounded-r-md border-l-0 border-gray-300 bg-neutral-600 px-2 font-semibold text-white shadow-sm">
+      <button id="search-btn" type="submit" class="sm:px4 rounded-r-md border-l-0 border-gray-300 bg-neutral-600 px-2 font-semibold text-white shadow-sm">
```
we are using nested group of sub divs now, so using div ids

Joydev commented on frontend/src/audios.html:
```diff
@@ -61,12 +61,24 @@
         </svg>
         <span class="hidden sm:inline">Search</span>
       </button>
+      <button
+        id="search-spinner"
+        type="button"
+        class="inline-flex items-center rounded-r-md bg-neutral-600 px-4 py-2 font-semibold leading-6 text-white"
+        disabled=""
+      >
```
done

Commit:
search-bar name updated to search-box

Commit:
update for search spinner

Commit:
updated top padding for spinner in small screen

northern-mindset committed:
Simplify loading indicator

northern-mindset committed:
Adjust layout of the error indicator

northern-mindset committed:
Rename the Error widget (from Status widget)

northern-mindset committed:
Remove redundant element manipulation

The error element is shown in the `render` event handler,
no need doing this in the `error` event handler.
